README.md
# Ambra-SDK

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
[![PyPI version](https://badge.fury.io/py/ambra-sdk.svg)](https://badge.fury.io/py/ambra-sdk)



---

Welcome to ambra-sdk library for intract with ambrahealth service and storage api. 


## Quickstart

```bash
pip install ambra-sdk
```

## Running

```python
from ambra_sdk.api import Api
from ambra_sdk.models import Study
from ambra_sdk.service.filtering import Filter, FilterCondition
from ambra_sdk.service.sorting import Sorter, SortingOrder

# Usually, URL has a form:
# url = https://ambrahealth_host/api/v3
# username and password - ambrahealth credentials.
api = Api.with_creds(url, username, password)
user_info = api.Session.user().get()

studies = api \
    .Study \
    .list() \
    .filter_by(
        Filter(
            'phi_namespace',
            FilterCondition.equals,
            user_info.namespace_id,
        ),
    ) \
    .only([Study.study_uid, Study.image_count]) \
    .sort_by(
        Sorter(
            'created',
            SortingOrder.ascending,
        ),
    ) \
    .all()

for study in studies:
    print(study.study_uid, study.image_count)
 
```

## License

Ambra-SDK is licensed under the terms of the Apache-2.0 License (see the file LICENSE).

## Read the docs

Documentation: https://dicomgrid.github.io/sdk-python/index.html


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [Unreleased]


## [3.22.4.0-1] - 2022-08-03
### Added
- Service public api support: LBL0022 v55.0 2022-08-03
- Storage public api support: LBL0038 v23.0 2022-08-03

## [3.22.3.0-1] - 2022-05-25
### Added
- Service public api support: LBL0022 v53.0 2021-04-05
- Storage public api support: LBL0038 v22.0 2022-05-18

## [3.22.1.0-1] - 2022-04-12
### Added
- Service public api support: LBL0022 v53.0 2021-04-05
- Storage public api support: LBL0038 v21.0 2022-04-05
- Files larger than 20 MB are now uploaded using multipart upload
- On regular upload to storage, the study_uid included as a query parameter

## [3.21.6.0-1] - 2021-08-11
### Added
- Cast request args to valid parameters
- Cast values in filters to valid ones
- Storage public api support: LBL0038 v16.0 2021-08-11
- Response text in unknown SDK exceptions

## [3.21.5.0-1] - 2021-06-30
### Added
- Asyncio !!!
- Service public api support: LBL0022 v47.0 2021-06-30
- Service models: LBL0022 v47.0 2021-06-30
- Rate limits
- Descriptive storage errors (new storage errors mechanism)
- Not json responses from v3services

## [3.21.4.0-1] - 2021-05-19
### Added
- Service public api updated (same veresion)
- Service models updated (same veresion)
- Storage public api support: LBL0038 v15.0 2021-05-19
- Support list in ERROR_TYPE v3services response
- Added is_ai argument for api.Storage.Study.Anonymize

### Fixed
- Removed dataclasses (python 3.6 support)
- Storage api args (used java classes)
- Deprecation warning messages
- Timeout, ws_timeout in anonymize_and_{wait,get} ignored
- Unicode symbols in docs (only for params description)

## [3.21.3.0-2] - 2021-04-16
### Fixed
- api.Namespace.settings settings -> ai_settings

## [3.21.3.0-1] - 2021-04-07
### Added
- Service public api support: LBL0022 v45.0 2021-04-07
- Service models: LBL0022 v45.0 2021-04-07
- Storage public api support: LBL0038 v14.0 2021-02-24
- api.Addon.Addon.Job.wait_completion method
- Job support for api.Storage.Study.delete_images
- Job support for api.Storage.Study.post_attachemnt
- Job support for api.Storage.Study.delete_attachemnt
- Job support for api.Storage.Study.merge
- Job support for api.Storage.Study.anonymize
- Job support for api.Storage.Study.clone
- Job support for api.Storage.Image.upload
- Job support for api.Storage.Image.wrap

## [3.21.2.0-1] - 2021-02-24
### Added
- api.Addon.Dicom.upload method
- api.Addon.Study.upload_dir method
- api.Addon.Study.upload_paths method
- api.Addon.Study.upload_dir_and_get method
- api.Addon.Study.upload_paths_and_get method
- api.Addon.Addon.Job.wait method
- api.Addon.Dicom.get method
- Session level setup headers default_headers
- Session level setup headers service_default_headers
- Session level setup headers storage_default_headers
- special_headers_for_login in api constructor

### Changed
- api.Addon.Study.upload_dicom method deprecated
- api.Addon.Study.upload method deprecated
- api.Addon.Study.upload_and_get method deprecated
- api.Addon.Study.wait_job method deprecated
- api.Addon.Study.dicom method deprecated
- Query.requst_params -> Query.request_args You can modify request arguments for specific service call

## [3.21.1.0-1] - 2021-01-13
### Added
- Service public api support: LBL0022 v44.0 2021-02-24
- Service models: LBL0022 v44.0 2021-02-24
- Storage public api support: LBL0038 v13.0 2021-01-13
- api.Addon.Study.wait_job method
- api.Addon.Study.wait_anonymize_and_wait method
- api.Addon.Study.wait_anonymize_and_get method
- Support filtering using datetime with timezones

## [3.20.8.0-2] - 2020-11-18
### Added
- Service public api support: LBL0022 v42.0 2020-12-02
- Service models: LBL0022 v42.0 2020-12-02
- New method api.Storage.Image.dicom_payload: return dicom payload
- New method api.Addon.Study.dicom: return pydicom object

## [3.20.7.0-1] - 2020-10-07
### Added
- Service public api support: LBL0022 v41.0 2020-10-07
- Service models: LBL0022 v41.0 2020-10-07
- Storage public api support: LBL0038 v11.0 2020-10-07

### Changed
- Using dashes instead of underscores in SDK header names (for example SDK_VERSION -> SDK-VERSION)

## [3.20.6.0-2] - 2020-08-26
### Fixed
- Error handling with list error_subtype

## [3.20.6.0-1] - 2020-08-26
### Added
- Documentation updates
- Service public api support: LBL0022 v40.0 2020-08-26
- Service models: LBL0022 v40.0 2020-08-26
- Storage public api support: LBL0038 v10.0 2020-08-26

### Changed
- Removed 502, 503 errors from default retry service params
- Update python-box dependence to 5.1.1
- Update requests dependence to 2.24.0

## [3.20.5.0-1] - 2020-07-15
### Added
- url, full_url, request_data for service query objects
- only_prepare argument (default is False) for all storage methods. If it is True, method return PreparedRequest object
- Examples of using retry mechanism for specific methods
- client_name argument in Api constructors
- SDK_VERSION, SDK_CLIENT_NAME in default session headers
- SDK version in documentation
- Service public api support: LBL0022 v39.0 2020-07-15
- Service models: LBL0022 v39.0 2020-07-15

### Changed
- Update to pydicom==2.0.0
- Change INFO log level in ws to DEBUG

### Fixed
- Fix usage example in README
- Fix retrying requests with new SID

## [3.20.4.0-1] - 2020-06-03
### Added
- Storage api support: LBL0038 v9.0 2020-06-03
- Service public api support: LBL0022 v38.0 2020-05-27
- Service models: LBL0022 v38.0 2020-05-27
- get_tags and tag_by_name methods in Storage.Study.json Box response
- get_tags and tag_by_name methods in Storage.Study.image_json Box response
- Add InconsistencyConfict(409) error Storage.Study.schema

### Changed
- Box response type in storage methods now is optional (default return Box)

## [3.20.3.0-5] - 2020-05-06
### Added
- Doctests in sphinx documentation
- Testing on various types of storage

### Changed

ambra_sdk/api/api.py
"""Ambra storage and service API."""

import logging
from threading import Lock
from time import monotonic, sleep
from typing import Callable, Optional, TypeVar

import requests
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util import Retry

from ambra_sdk.addon.addon import Addon
from ambra_sdk.api.base_api import BaseApi
from ambra_sdk.clear_params import clear_params
from ambra_sdk.exceptions.service import AuthorizationRequired
from ambra_sdk.exceptions.storage import AccessDenied
from ambra_sdk.request_args import RequestArgs
from ambra_sdk.service.entrypoints import (
    Account,
    Activity,
    Analytics,
    Annotation,
    Anonymization,
    Appointment,
    Audit,
    Case,
    Customcode,
    Customfield,
    Destination,
    Dicomdata,
    Dictionary,
    Filter,
    Group,
    Help,
    Hl7,
    Keyimage,
    Link,
    Location,
    Meeting,
    Message,
    Namespace,
    Node,
    Npi,
    Order,
    Patient,
    Purge,
    Qctask,
    Query,
    Radreport,
    Radreportmacro,
    Report,
    Role,
    Route,
    Rsna,
    Scanner,
    Session,
    Setting,
    Site,
    Study,
    Tag,
    Terminology,
    Training,
    User,
    Validate,
    Webhook,
)
from ambra_sdk.storage.storage import Storage

logger = logging.getLogger(__name__)


class Api(BaseApi):
    """Ambra API.

    Example:
    >>> from ambra_sdk.api import Api
    >>>
    >>> api = Api.with_creds(
    ...     url='https://ambrahealth_host/api/v3',
    ...     username='username',
    ...     password='password',
    ...     client_name='Some name (ex: Hospital-ABC)',
    ... )

    Using some special parameters (In most cases you don't need this!):

    >>> from ambra_sdk.api import Api
    >>> from ambra_sdk.api.base_api import RateLimit, RateLimits
    >>>
    >>> MY_RLS = RateLimits(
    ...        default=RateLimit(3, 2),
    ...        get_limit=RateLimit(4, 2),
    ...        special={'special_url': RateLimit(5, 2)},
    ... )
    >>>
    >>> api = Api.with_creds(
    ...     url='https://ambrahealth_host/api/v3',
    ...     username='username',
    ...     password='password',
    ...     client_name='Some name (ex: Hospital-ABC)',
    ...     special_headers_for_login={'Special-H': 'ABC'},
    ...     rate_limits=MY_RLS,
    ...     autocast_arguments=False,
    ... )
    """

    backend = 'REQUESTS'

    def __init__(self, *args, **kwargs):
        """Init api.

        :param args: args
        :param kwargs: kwargs
        """
        super().__init__(*args, **kwargs)

        self._service_session: Optional[requests.Session] = None
        self._storage_session: Optional[requests.Session] = None
        self._init_request_params()

        # For rate limits
        self._rate_limits_lock = Lock()
        self._last_request_time = None
        self._last_call_period = None

        # Init services api
        self._init_service_entrypoints()

        # Init storage api namespace
        self.Storage = Storage(self)

        # Init addon namespace
        self.Addon = Addon(self)

    @property
    def service_session(self) -> requests.Session:
        """Service session.

        :return: service session
        """
        if self._service_session is None:
            self._service_session = requests.Session()
            retries = Retry(**self.service_retry_params)
            adapter = HTTPAdapter(max_retries=retries)
            self._service_session.mount('http://', adapter)
            self._service_session.mount('https://', adapter)
            self._service_session.headers.update(self._default_headers)
            self._service_session.headers.update(self._service_default_headers)
        return self._service_session

    @property
    def storage_session(self) -> requests.Session:
        """Storage session.

        :return: storage session
        """
        if self._storage_session is None:
            self._storage_session = requests.Session()
            retries = Retry(**self.storage_retry_params)
            adapter = HTTPAdapter(max_retries=retries)
            self._storage_session.mount('http://', adapter)
            self._storage_session.mount('https://', adapter)
            self._storage_session.headers.update(self._default_headers)
            self._storage_session.headers.update(self._storage_default_headers)
        return self._storage_session

    def storage_get(
        self,
        url: str,
        required_sid: bool,
        **kwargs,
    ) -> requests.Response:
        """Get from storage.

        :param url: url
        :param required_sid: is this method required sid
        :param kwargs: request arguments
        :return: response obj
        """
        kwargs = self._prepare_storage_request_args(
            required_sid=required_sid,
            **kwargs,
        )
        logger.info(
            'Storage get: %s. Params: %s',
            url,
            str(clear_params(kwargs.get('params', {}))),
        )
        return self.storage_session.get(url=url, **kwargs)

    def storage_delete(
        self,
        url: str,
        required_sid: bool,
        **kwargs,
    ) -> requests.Response:
        """Delete from storage.

        :param url: url
        :param required_sid: is this method required sid
        :param kwargs: request arguments
        :return: response obj
        """
        kwargs = self._prepare_storage_request_args(
            required_sid=required_sid,
            **kwargs,
        )
        logger.info(
            'Storage delete: %s. Params: %s',
            url,
            str(clear_params(kwargs.get('params', {}))),
        )
        return self.storage_session.delete(url=url, **kwargs)

    def storage_post(
        self,
        url: str,
        required_sid: bool,
        **kwargs,
    ) -> requests.Response:
        """Post to storage.

        :param url: url
        :param required_sid: is this method required sid
        :param kwargs: request arguments
        :return: response obj
        """
        kwargs = self._prepare_storage_request_args(
            required_sid=required_sid,
            **kwargs,
        )
        logger.info(
            'Storage post: %s. Params: %s',
            url,
            str(clear_params(kwargs.get('params', {}))),
        )
        return self.storage_session.post(url=url, **kwargs)

    def service_request(
        self,
        request_args: RequestArgs,
        required_sid: bool,
    ) -> requests.Response:
        """Post data to url.

        :param request_args: request args
        :param required_sid: is this method required sid
        :return: response
        """
        self._wait_for_service_request(request_args.url)
        return self._service_request_without_rate_limits(
            request_args=request_args,
            required_sid=required_sid,
        )

    def service_post(
        self,
        url: str,
        required_sid: bool,
        **kwargs,
    ) -> requests.Response:
        """Post data to url.

        :param url: method url
        :param required_sid: is this method required sid
        :param kwargs: request arguments
        :return: response
        """
        full_url = self.service_full_url(url)
        request_args = RequestArgs(
            method='POST',
            url=url,
            full_url=full_url,
            **kwargs,
        )
        return self.service_request(
            request_args=request_args,
            required_sid=required_sid,
        )

    def get_sid(self) -> str:
        """Get or create new sid.

        :return: sid
        """
        if self._sid is None:
            return self.get_new_sid()
        return self._sid

    @property
    def sid(self) -> str:
        """Get or create new sid property.

        :return: sid
        """
        return self.get_sid()

    def logout(self):
        """Logout."""
        if self._sid:
            self.Session.logout().get()
            self._sid = None
        if self._storage_session:
            self._storage_session.close()
        if self._service_session:
            self._service_session.close()

    def get_new_sid(self) -> str:
        """Get new sid.

        :raises RuntimeError: Missed credentials
        :return: sid
        """
        if self._creds is None:
            raise RuntimeError('Missed credentials')
        new_sid: str = self.Session.get_sid(
            self._creds.username,
            self._creds.password,
            special_headers_for_login=self._special_headers_for_login,
        )
        self._sid = new_sid
        return new_sid

    FN_RETURN_TYPE = TypeVar('FN_RETURN_TYPE')

    def retry_with_new_sid(
        self,
        fn: Callable[..., FN_RETURN_TYPE],
    ) -> FN_RETURN_TYPE:
        """Retry with new sid.

        :param fn: callable method
        :return: fn result
        """
        try:
            return fn()
        except (AuthorizationRequired, AccessDenied):
            self.get_new_sid()
            return fn()

    def _wait_for_service_request(self, url):
        if self._rate_limits:
            with self._rate_limits_lock:
                call_period = self._rate_limits.call_period(url)
                now = monotonic()
                if self._last_request_time is None:
                    # This is a first run
                    self._last_request_time = now
                    self._last_call_period = call_period
                    return

                wait_time = self._last_request_time + \
                    self._last_call_period - now
                if wait_time > 0:
                    logger.info('Sleep %s due to rate limits', wait_time)
                    sleep(wait_time)
                self._last_request_time = monotonic()
                self._last_call_period = call_period

    def _service_request_without_rate_limits(
        self,
        request_args: RequestArgs,
        required_sid: bool,
    ) -> requests.Response:
        """Post data to url.

        :param request_args: request args
        :param required_sid: is this method required sid
        :return: response
        """
        if required_sid is True:
            request_data = request_args.data or {}
            request_data['sid'] = self.sid
            request_args.data = request_data  # NOQA:WPS110
        logger.info(
            'Service post: %s. Params: %s',
            request_args.url,
            str(clear_params(request_args.data)),
        )
        return self.service_session.request(
            method=request_args.method,
            url=request_args.full_url,
            **request_args.dict_optional_args(
                self._autocast_arguments,
            ),
        )

    def _prepare_storage_request_args(
        self,
        required_sid: bool,
        **kwargs,
    ):
        """Prepare storage request kwargs args.

        :param required_sid: required sid
        :param kwargs: kwargs
        :return: kwargs
        """
        if required_sid:
            # Sid passed always in url params (?sid=...)
            request_params = kwargs.pop('params')
            # Get or create new sid
            request_params['sid'] = self.get_sid()
            kwargs['params'] = request_params
        return kwargs

    def _init_request_params(self):
        method_whitelist = [
            'HEAD',
            'TRACE',
            'GET',
            'PUT',
            'OPTIONS',
            'DELETE',
            'POST',
        ]
        # https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#module-urllib3.util.retry
        self.service_retry_params = {
            'total': 10,
            'connect': 5,
            'read': 5,
            'status': 5,
            'status_forcelist': [503, 504],
            'backoff_factor': 0.1,
            'method_whitelist': method_whitelist,
        }
        # Merge studies
        # /study/{namespace}/{studyUid}/merge?sid={sid}&secondary_study_uid={secondary_study_uid}&delete_secondary_study={0,1}
        # Can return 500 and its ok .....
        self.storage_retry_params = {
            'total': 10,
            'connect': 5,
            'read': 5,
            'status': 5,
            'status_forcelist': [502, 503, 504],
            'backoff_factor': 0.1,
            'method_whitelist': method_whitelist,
        }

    def _init_service_entrypoints(self):
        """Init service entrypoint namespaces."""
        self.Account = Account(self)
        self.Activity = Activity(self)
        self.Analytics = Analytics(self)
        self.Annotation = Annotation(self)
        self.Appointment = Appointment(self)
        self.Audit = Audit(self)
        self.Case = Case(self)
        self.Customcode = Customcode(self)
        self.Customfield = Customfield(self)
        self.Destination = Destination(self)
        self.Dicomdata = Dicomdata(self)
        self.Dictionary = Dictionary(self)
        self.Filter = Filter(self)
        self.Group = Group(self)
        self.Help = Help(self)
        self.Hl7 = Hl7(self)
        self.Keyimage = Keyimage(self)
        self.Link = Link(self)
        self.Location = Location(self)
        self.Meeting = Meeting(self)
        self.Message = Message(self)
        self.Namespace = Namespace(self)
        self.Node = Node(self)
        self.Npi = Npi(self)
        self.Order = Order(self)
        self.Patient = Patient(self)
        self.Purge = Purge(self)
        self.Radreportmacro = Radreportmacro(self)
        self.Radreport = Radreport(self)
        self.Report = Report(self)
        self.Role = Role(self)
        self.Route = Route(self)
        self.Rsna = Rsna(self)
        self.Session = Session(self)
        self.Setting = Setting(self)
        self.Study = Study(self)
        self.Tag = Tag(self)
        self.Terminology = Terminology(self)
        self.Training = Training(self)
        self.User = User(self)
        self.Validate = Validate(self)
        self.Webhook = Webhook(self)
        self.Query = Query(self)
        self.Scanner = Scanner(self)
        self.Site = Site(self)
        self.Anonymization = Anonymization(self)
        self.Qctask = Qctask(self)
